### PR TITLE
Resolve MTA properties of a module, given env-variables and/or .env file

### PR DESCRIPTION
**Pull request by ShimiT**
Before submitting a pull request, Please make sure the following:

## Description

Please explain the changes you made.

Link to design (if applicable)

### Checklist
- [x] Code compiles correctly
- [ ] Relevant tests were added (unit / contract / integration)
- [ ] Relevant logs were added
- [ ] Formating and linting run locally successfully
- [x] All tests passing
- [ ] UA review
- [ ] Design is documented
- [ ] Extended the README / documentation, if necessary
- [ ] Open source is approved

### CODE REVIEW

File: cmd/mtaresolver.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+func (m *MTAResolver) ResolveProperies(module *mta.Module) {
+
+	if m.Parameters == nil {
+		m.Parameters = map[string]interface{}{}
+	}
+
+	//add env variables
+	for _, val := range os.Environ() {
+		pos := strings.Index(val, "=")
```
Comment (allaVolkov) [inline cmd/mtaresolver.go]: I should suggest to use strings.SplitN function

File: cmd/mtaresolver.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+	}
+}
+
+func (m *MTAResolver) addValueToContext(key, value string) {
+	//if the key has format of "module/key", or "resource/key" writes the value to the module's context
+	slashPos := strings.Index(key, "/")
+	if slashPos > 0 {
+		modName := key[:slashPos]
+		key = key[slashPos+1:]
+		modContext, ok := m.context.modules[modName]
```
Comment (allaVolkov) [inline cmd/mtaresolver.go]: it's worth to rename modContext to (for example) context

File: cmd/mtaresolver.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+	if sourceModule != nil {
+		paramVal := sourceModule.Parameters[paramName]
+		if paramVal != nil {
+			return paramVal.(string)
+		}
+		//defaults to context's module params:
+		paramValStr, ok := m.context.modules[sourceModule.Name][paramName]
+		if ok {
+			return paramValStr
+		}
```
Comment (allaVolkov) [inline cmd/mtaresolver.go]: isn't search on m.context.resources relevant?

File: cmd/json.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+func ConvertToJsonSafe(val interface{}) interface{} {
+	switch v := val.(type) {
+	case map[interface{}]interface{}:
+		//valMap :=
+		res := map[string]interface{}{}
+		for k, v := range v {
+			switch v2 := v.(type) {
+			case map[interface{}]interface{}, []interface{}:
```
Comment (allaVolkov) [inline cmd/json.go]: isn't it enough res[k] = ConvertToJsonSafe(v)?

File: cmd/errors.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/errors.go]: removed file

File: cmd/json.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/DisposaBoy/JsonConfigReader"
+)
+
+func ParseJSONWithComments(input []byte, obj interface{}) error {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/json.go]: remove file

File: cmd/errors.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package commands
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/errors.go]: remove file

File: cmd/json.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	r := JsonConfigReader.New(bytes.NewReader(input))
+	return json.NewDecoder(r).Decode(obj)
+}
+
+func ConvertToJsonSafe(val interface{}) interface{} {
+	switch v := val.(type) {
+	case map[interface{}]interface{}:
+		//valMap :=
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/json.go]: done

File: cmd/mtaresolver.go [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+const RESOURCE_TYPE = 1
+const MODULE_TYPE = 2
+
+const VARIABLE_PREFIX = "~"
+const PLACEHOLDER_PREFIX = "$"
+const TEMPLATE_PREIX = "^"
+
+type MTASource struct {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: done

File: cmd/resolver.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+var resolveMtaCmd = &cobra.Command{
+	Use:   "resolve",
+	Short: "Resolve variables and placeholders in an MTA file",
+	Long: `MTA file typically contains variables in the form ~{var-name} and placeholders in the form ${placeholder}, 
+resolve command print to stdout the MTA fil contents with as much as possible variables and placeholders replaced 
+with concrete values, based on environment variables provided and .env files in the modules' folders`,
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		logs.Logger.Info("Resolve MTA")
+		err := resolve(resolvePath)
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/resolver.go]: changed

File: cmd/mtaresolver.go (hunk 2 above)
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: changed to modulesContext. context may confuse with m.context

File: cmd/mtaresolver.go (hunk 1 above)
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: I prefer this way - more clear, as I always want the first

File: cmd/platform_parameters.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+type ResolveContext struct {
+	global    map[string]string
+	modules   map[string]map[string]string
+	resources map[string]map[string]string
+}
+
+func (m *MTAResolver) addServiceNames(module *mta.Module) {
+
+	for _, resource := range m.Resources {
+		if m.context.resources[resource.Name] == nil {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/platform_parameters.go]: removed

File: cmd/mtaresolver.go [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+		if pos >= 0 {
+			varValue = m.getVariableValue(sourceModule, requires, variableName)
+		}
+	}
+
+	return value
+}
+
+func convertToString(valueObj interface{}) (string, bool) {
+	switch valueObj.(type) {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: switch is better in type checking, I also now use the typed value instead of another cast

File: cmd/json.go (hunk 4 above)
Comment (ariel-bentu) [reply cmd/json.go]: probably 			res[fmt.Sprint(k)] = convertToJSONSafe(v) will be ok - changed

File: cmd/mtaresolver.go (hunk 3 above)
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: No. within a module, you search in the module, or in its required resource or module, or globally. but not arbitrary resource which is not required...

File: cmd/platform_parameters.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+		//try to find the service-name in VCAP_SERVICES
+		serviceName := m.getServiceInstanceFromEnv(resource)
+		if len(serviceName) > 0 {
+			resCtx["service-name"] = serviceName
+		}
+	}
+}
+
+func (m *MTAResolver) getServiceInstanceFromEnv(resource *mta.Resource) string {
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/platform_parameters.go]: done

File: cmd/mtaresolver.go [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+
+func (m *MTAResolver) getVariableValue(sourceModule *mta.Module, requires *mta.Requires, variableName string) interface{} {
+	var providerName string
+	if requires == nil {
+		slashPos := strings.Index(variableName, "/")
+		if slashPos > 0 {
+			providerName = variableName[:slashPos]
+			variableName = variableName[slashPos+1:]
+		} else {
+			panic(fmt.Errorf("Cannot resolve value for variable ~{%s}. Missing requires prefix", variableName))
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: removed

File: cmd/mtaresolver.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: removed

File: cmd/mtaresolver.go [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+func convertToString(valueObj interface{}) (string, bool) {
+	switch valueObj.(type) {
+	case string:
+		return valueObj.(string), false
+	}
+	valueBytes, err := json.Marshal(ConvertToJsonSafe(valueObj))
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: removed

File: mta/mta.go [hunk 16]
```diff
@@ -52,13 +52,14 @@ func (mta *MTA) GetResourceByName(name string) (*Resource, error) {
 // Unmarshal returns a reference to the MTA object from a byte array.
 func Unmarshal(content []byte) (*MTA, error) {
 	dec := yaml.NewDecoder(bytes.NewReader(content))
-	dec.KnownFields(true)
+	//dec.KnownFields(true)
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply mta/mta.go]: Shimi knows about this. he got from me an example yaml that v3 fails to parse - please decide what to do

File: cmd/mtaresolver.go [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"os"
+	"strings"
+
+	"github.com/SAP/cloud-mta/mta"
```
[the remark this answers is not in the document]
Comment (ariel-bentu) [reply cmd/mtaresolver.go]: alphabetically? done

[the remark this answers is not in the document]
Comment (allaVolkov) [reply cmd/mtaresolver.go]: yes plus grouped in the following order: go libs, github (not SAP), github (SAP).

**Comment by allaVolkov** [comment]
@ariel-bentu Please review unit tests added.

**Pull request merged**